### PR TITLE
First version of Rest Implementation in Secure Sample Plugin

### PR DESCRIPTION
**Pull request by sudhirappaji**
First Version of Rest Implementation in sample secure plugin.
It  contains the following:
1.	Rest Framework
2.	Empty ManagerPolicyElement for GET/PUSH/PUT/DELETE

This can be tested using Fiddler for sending REST requests:

a.	GET with https://localhost:8090/testsmplugin/sm/server/v1/policies
b.	GET with https://localhost:8090/testsmplugin/sm/server/v1/policies/{id}
c.	POST with https://localhost:8090/testsmplugin/sm/server/v1/policies
Request body to contain:
{
 "id":1,
"name":"anyname”
}
d.	PUT OR DELETE with https://localhost:8090/testsmplugin/sm/server/v1/policies/{id}
{
 "id":1,
"name":"anyname”}

For all the above Fiddler menu to have below configuration:
User-Agent: Fiddler
Accept: application/json
Content-Type: application/json
Authorization: Basic YWRtaW46YWRtaW4xMjM=
Host: localhost:8090
Content-Length: 41

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -116,6 +116,57 @@
...
 		</dependency>
 
+	 	<dependency>
+			<groupId>com.fasterxml.jackson.jaxrs</groupId>
+			<artifactId>jackson-jaxrs-json-provider</artifactId>
+		  	<version>2.8.5</version>
+		</dependency>
+		<dependency>
+			<groupId>com.fasterxml.jackson.core</groupId>
+    		<artifactId>jackson-core</artifactId>
```
Comment (emanoelxavier) [inline pom.xml]: Fix indentation

File: src/main/java/org/osc/manager/ism/entities/PolicyElements.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+package org.osc.manager.ism.entities;
+
+
+
+
+public class PolicyElements {
```
Comment (emanoelxavier) [inline src/main/java/org/osc/manager/ism/entities/PolicyElements.java]: Rename PolicyElements to PolicyEntity. We can use the suffix *Entity for all the classes inside this package for consistency. 
Also I think that we will no longer need the classes inside the package ism.model. We can probably just have the *.entities package now to simplify and just make sure those classes implement the *Element interfaces. In other words, delete PolicyListElement and make this class implement the SDK interface ManagerPolicyElement. @arvindn05  do you agree?

File: src/main/java/org/osc/manager/rest/server/SecureManagerServerRestConstants.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+package org.osc.manager.rest.server;
+
+public interface SecureManagerServerRestConstants {
+    // Because we started with versions in the URL, we will continue to have v1 in the URL
+    String SERVER_API_PATH_PREFIX = "/sm/server/v1";
```
Comment (emanoelxavier) [inline src/main/java/org/osc/manager/rest/server/SecureManagerServerRestConstants.java]: Since you have /sm/ here change the url from /testsmplugin/sm/.... to /testplugin/sm/ and for the sdn controller it will be /testplugin/sdn/....

File: src/main/java/org/osc/manager/rest/server/SecureManagerServletDelegate.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+        HTTP_WHITEBOARD_SERVLET_NAME + "=" + "OSC-API", HTTP_WHITEBOARD_SERVLET_PATTERN + "=/testsmplugin/*",
+        HTTP_WHITEBOARD_CONTEXT_SELECT + "=(" + HTTP_WHITEBOARD_CONTEXT_NAME + "=" + "OSC-API"
+        + ")",        
+        HTTP_WHITEBOARD_TARGET + "=(" + "org.apache.felix.http.name" + "=" + "OSC-API" + ")"
+        
+ })
+
+
+public class SecureManagerServletDelegate extends ResourceConfig implements Servlet {
```
Comment (emanoelxavier) [inline src/main/java/org/osc/manager/rest/server/SecureManagerServletDelegate.java]: SecureManagerServletDelegate -> SecurityManagerServletDelegate

File: src/main/java/org/osc/manager/rest/server/SecureManagerServletDelegate.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+import org.osc.manager.rest.server.api.PolicyApis;
+//import org.osc.core.broker.rest.server.ApiServletContext;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import com.fasterxml.jackson.jaxrs.json.JacksonJaxbJsonProvider;
+
+@Component(name = "sm.servlet", service = Servlet.class, property = {
+
+        HTTP_WHITEBOARD_SERVLET_NAME + "=" + "OSC-API", HTTP_WHITEBOARD_SERVLET_PATTERN + "=/testsmplugin/*",
```
Comment (emanoelxavier) [inline src/main/java/org/osc/manager/rest/server/SecureManagerServletDelegate.java]: testsmplugin -> tstplugin. If you must keep testsm here to disambiguate from the sdn controller then remove the other /sm/ bit from the url

File: src/main/java/org/osc/manager/rest/server/SecureManagerServerRestConstants.java (hunk 3 above)
Comment (emanoelxavier) [inline src/main/java/org/osc/manager/rest/server/SecureManagerServerRestConstants.java]: we probably dont need /server/v1 on these urls. just go for testsmplugin/policies or testplugin/sm/policies

File: src/main/java/org/osc/manager/rest/server/SecureManagerServletDelegate.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+   
+   @Activate
+   void activate() {
+       // Json feature
+	   super.register(JacksonJaxbJsonProvider.class);
+
+       //Properties
+       super.property(ServerProperties.FEATURE_AUTO_DISCOVERY_DISABLE, true);
+
+       // agent & server apis
```
Comment (emanoelxavier) [inline src/main/java/org/osc/manager/rest/server/SecureManagerServletDelegate.java]: delete comment

File: src/main/java/org/osc/manager/rest/server/SecureManagerServletDelegate.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+ })
+
+
+public class SecureManagerServletDelegate extends ResourceConfig implements Servlet {
+    static final long serialVersionUID = 1L;
+
+   @Reference
+   private PolicyApis policiesApis;
+
+
```
Comment (emanoelxavier) [inline src/main/java/org/osc/manager/rest/server/SecureManagerServletDelegate.java]: remove extra lines

File: src/main/java/org/osc/manager/rest/server/api/PolicyApis.java [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.HttpHeaders;
+import javax.ws.rs.core.MediaType;
+
+import org.apache.log4j.Logger;
+import org.osc.manager.rest.server.SecureManagerServerRestConstants;
+import org.osc.manager.ism.entities.PolicyElements;
+import org.osgi.service.component.annotations.Component;
+
+
```
Comment (emanoelxavier) [inline src/main/java/org/osc/manager/rest/server/api/PolicyApis.java]: remove extra lines. Please look everywhere

File: src/main/java/org/osc/manager/rest/server/api/PolicyApis.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+	
+	private static final Logger logger = Logger.getLogger(PolicyApis.class);
+
+    /**
+     * Creates the ISM policies
+     *
+     * @return policy
+     */
+    @POST
+    public PolicyElements createIsmPolicies(@Context HttpHeaders headers ,PolicyElements elements){
```
Comment (emanoelxavier) [inline src/main/java/org/osc/manager/rest/server/api/PolicyApis.java]: wrong spaces around `,`

Comment (emanoelxavier) [inline src/main/java/org/osc/manager/rest/server/api/PolicyApis.java]: createIsmPolicies -> createPolicy since it creates just one. You can use the same convention for consistency: updatePolicy, getPolicyIds, getPolicy, deletePolicy

File: src/main/java/org/osc/manager/rest/server/api/PolicyApis.java [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+    	
+    	logger.info("Creating Policies Elements...");
+       
+    	// TODO DB Create
```
Comment (emanoelxavier) [inline src/main/java/org/osc/manager/rest/server/api/PolicyApis.java]: Whenever add a TODO always include the name of the person who will address the TODO, i.e.: // TODO sudhir: Add db calls here. Also update other places.

File: src/main/java/org/osc/manager/rest/server/api/PolicyApis.java [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+    }
+
+    
+    /**
+     * Lists the ISM Policy
+     *
+     * @return - Policy List
+     */
+	@GET
+    public List<PolicyElements> getPolicyList(@Context HttpHeaders headers) {
```
Comment (emanoelxavier) [inline src/main/java/org/osc/manager/rest/server/api/PolicyApis.java]: The get should return a list of their ids, List<String> instead of the entire element

File: src/main/java/org/osc/manager/rest/server/api/PolicyApis.java [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+	@GET
+    public PolicyElements getPolicy(@Context HttpHeaders headers,@PathParam("policyId") Long policyId) {
+		
+    	 logger.info("getting Policy " + policyId);
+        
+		// TODO retrieve from DB 
+		
+		return null;
+    }
+
```
Comment (emanoelxavier) [inline src/main/java/org/osc/manager/rest/server/api/PolicyApis.java]: observe empty lines

File: pom.xml (hunk 1 above)
Comment (sudhirappaji) [inline pom.xml]: OK will fix this

File: src/main/java/org/osc/manager/ism/entities/PolicyElements.java (hunk 2 above)
Comment (sudhirappaji) [inline src/main/java/org/osc/manager/ism/entities/PolicyElements.java]: OK, shall rename PolicyElement to PolicyEntity. For renaming other classes to suffix *Entity shall be done as part of respective user stories and not part of this change.

Comment (sudhirappaji) [inline src/main/java/org/osc/manager/ism/entities/PolicyElements.java]: Regarding, deleting PolicyListElement and extending PolicyEntity to implement ManagerPolicyElement, I still see the need to have PolicyEntity and PolicyListElement as PolicyEntity is used for  DB updates.
Also I would like investigate this as part of another user story separately and not part of this request. Let me know what you think? or we discuss further offline.

File: src/main/java/org/osc/manager/rest/server/SecureManagerServerRestConstants.java (hunk 3 above)
Comment (sudhirappaji) [inline src/main/java/org/osc/manager/rest/server/SecureManagerServerRestConstants.java]: will change as proposed to /testplugin/sm, /testplugin/sdn/...

Comment (sudhirappaji) [inline src/main/java/org/osc/manager/rest/server/SecureManagerServerRestConstants.java]: OK, as indicated in earlier comment

File: src/main/java/org/osc/manager/rest/server/SecureManagerServletDelegate.java (hunk 5 above)
Comment (sudhirappaji) [inline src/main/java/org/osc/manager/rest/server/SecureManagerServletDelegate.java]: OK will change to testplugin

File: src/main/java/org/osc/manager/rest/server/SecureManagerServletDelegate.java (hunk 4 above)
Comment (sudhirappaji) [inline src/main/java/org/osc/manager/rest/server/SecureManagerServletDelegate.java]: OK will change to SecurityManagerServletDelegate

File: src/main/java/org/osc/manager/rest/server/SecureManagerServletDelegate.java (hunk 6 above)
Comment (sudhirappaji) [inline src/main/java/org/osc/manager/rest/server/SecureManagerServletDelegate.java]: OK

File: src/main/java/org/osc/manager/rest/server/SecureManagerServletDelegate.java (hunk 7 above)
Comment (sudhirappaji) [inline src/main/java/org/osc/manager/rest/server/SecureManagerServletDelegate.java]: OK

File: src/main/java/org/osc/manager/rest/server/api/PolicyApis.java (hunk 8 above)
Comment (sudhirappaji) [inline src/main/java/org/osc/manager/rest/server/api/PolicyApis.java]: ok

File: src/main/java/org/osc/manager/rest/server/api/PolicyApis.java (hunk 9 above)
Comment (sudhirappaji) [inline src/main/java/org/osc/manager/rest/server/api/PolicyApis.java]: ok

Comment (sudhirappaji) [inline src/main/java/org/osc/manager/rest/server/api/PolicyApis.java]: ok

File: src/main/java/org/osc/manager/rest/server/api/PolicyApis.java (hunk 10 above)
Comment (sudhirappaji) [inline src/main/java/org/osc/manager/rest/server/api/PolicyApis.java]: ok

File: src/main/java/org/osc/manager/rest/server/api/PolicyApis.java (hunk 11 above)
Comment (sudhirappaji) [inline src/main/java/org/osc/manager/rest/server/api/PolicyApis.java]: ok

File: src/main/java/org/osc/manager/rest/server/api/PolicyApis.java (hunk 12 above)
Comment (sudhirappaji) [inline src/main/java/org/osc/manager/rest/server/api/PolicyApis.java]: ok

File: src/main/java/org/osc/manager/ism/entities/PolicyElements.java (hunk 2 above)
Comment (emanoelxavier) [inline src/main/java/org/osc/manager/ism/entities/PolicyElements.java]: With PolicyEntity in place and implementing ManagerPolicyElement you should no longer need PolicyListElement, it should be quick to replace all references to PolicyListElement with PolicyEntity and delete PolicyListElement.  If you are tracking using the *Entities in the SDK interfaces implementation with different user stories you can do that later but I want to make sure you agree that all the *Entities will implement their corresponding *Element interfaces and there wont any other classes implementing those interfaces like Policy or PolicyListElement when the feature is complete. Also for you to consider an entire entity ready, like Policy, it should be: persisted in the db; exposed with those crud apis in the rest; **and used by the plugin on their SDK interface implementation (this will make OSC also use the persistence mechanism);**

Comment (arvindn05) [inline src/main/java/org/osc/manager/ism/entities/PolicyElements.java]: I agree. For the most part, the model classes might go away. We might still need some model classes to be present in case there is no direct mapping possible between the *Element interfaces and the entities.

Comment (sudhirappaji) [inline src/main/java/org/osc/manager/ism/entities/PolicyElements.java]: OK and in this version I shall make PolicyEntity implement  ManagerPolicyElement. The plugin cleanup shall be in the subsequent user stories which we had discussed earlier.

File: src/main/java/org/osc/manager/rest/server/SecurityManagerServerRestConstants.java [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+package org.osc.manager.rest.server;
+
+public interface SecurityManagerServerRestConstants {
+    // Because we started with versions in the URL, we will continue to have v1 in the URL
```
Comment (emanoelxavier) [inline src/main/java/org/osc/manager/rest/server/SecurityManagerServerRestConstants.java]: delete comment

File: src/main/java/org/osc/manager/rest/server/api/PolicyApis.java [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+    private static final Logger logger = Logger.getLogger(PolicyApis.class);
+
+    /**
+     * Creates the Policy
+     *
+     * @return policy
+     */
+    @POST
+    public PolicyEntity createPolicy(@Context HttpHeaders headers ,PolicyEntity entity) {
```
Comment (emanoelxavier) [inline src/main/java/org/osc/manager/rest/server/api/PolicyApis.java]: fix spaces around the comma (there is a eclipse setting you can turn on to fix this automatically upon saving)

Comment (emanoelxavier) [inline src/main/java/org/osc/manager/rest/server/api/PolicyApis.java]: you can probably also remove the headers parameter since the api implementation is not using it. make sure to try it out without this parameter and if not needed remove.

File: src/main/java/org/osc/manager/rest/server/api/PolicyApis.java [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    /**
+     * Updates the Policy
+     *
+     * @return - updated policy
+     */
+    @Path("/{policyId}")
+    @PUT
+    public PolicyEntity updatePolicy(@Context HttpHeaders headers ,@PathParam("policyId") Long policyId,PolicyEntity entity) {
```
Comment (emanoelxavier) [inline src/main/java/org/osc/manager/rest/server/api/PolicyApis.java]: spaces around the commas

File: src/main/java/org/osc/manager/rest/server/SecurityManagerServerRestConstants.java (hunk 13 above)
Comment (sudhirappaji) [inline src/main/java/org/osc/manager/rest/server/SecurityManagerServerRestConstants.java]: OK

File: src/main/java/org/osc/manager/rest/server/api/PolicyApis.java (hunk 14 above)
Comment (sudhirappaji) [inline src/main/java/org/osc/manager/rest/server/api/PolicyApis.java]: OK and have enabled the eclipse settings now.

Comment (sudhirappaji) [inline src/main/java/org/osc/manager/rest/server/api/PolicyApis.java]: OK, removed and it works. In addition, will make changes to PolicyEntity to reflect MC to multiple domains, domains to multiple policies. I plan to add two more members in PolicyEntity ==>ApplianceManagerConnectorElement applianceManagerConnector; and the DomainElement domain;
Also to reflect the above changes in PolicyEntity, PolicyApis content will go for a change. This should get reflected as part of this fix.

File: src/main/java/org/osc/manager/rest/server/api/PolicyApis.java (hunk 15 above)
Comment (sudhirappaji) [inline src/main/java/org/osc/manager/rest/server/api/PolicyApis.java]: OK, in addition will make changes to reflects  based on domainId and PolicyId.

File: src/main/java/org/osc/manager/rest/server/api/PolicyApis.java [hunk 16]
```diff
@@ -44,70 +42,71 @@
...
     }
 
     /**
-     * Updates the Policy
+     * Updates the Policy for a given domain
      *
      * @return - updated policy
      */
-    @Path("/{policyId}")
+    @Path("/{domainId}")
```
Comment (emanoelxavier) [inline src/main/java/org/osc/manager/rest/server/api/PolicyApis.java]: actually the policy id should also be in the url, typically for update, delete and get the targeted object id should be in the url as well. So the resource path should be: /domains/{domainid}/policies/{policyid} . See this example https://github.com/opensecuritycontroller/osc-core/blob/master/osc-rest-server/src/main/java/org/osc/core/broker/rest/server/api/VirtualizationConnectorApis.java#L343 where the security groups resource is under the virtualization connectors resource and follow the same pattern for this.

Comment (sudhirappaji) [inline src/main/java/org/osc/manager/rest/server/api/PolicyApis.java]: OK and I have fixed it.

File: src/main/java/org/osc/manager/rest/server/api/PolicyApis.java [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.MediaType;
+
+import org.apache.log4j.Logger;
+import org.osc.manager.ism.entities.PolicyEntity;
+import org.osc.manager.rest.server.SecurityManagerServerRestConstants;
+import org.osgi.service.component.annotations.Component;
+
+@Component(service = PolicyApis.class)
+@Path(SecurityManagerServerRestConstants.SERVER_API_PATH_PREFIX + "/policy")
```
Comment (emanoelxavier) [inline src/main/java/org/osc/manager/rest/server/api/PolicyApis.java]: Looks like this url will end up being /policy/domains/domainid/{did}/policies/{policyid}, it should be just /domains/{did}/policies/{policyid}, without the initial  '/policy/'.  For that you may need to change the route here to be actually **/domains**. Observe that it will be the same route as the DomainApis when you have that in the future , which means at least two classes DomainApis and PolicyApis will have the same route, if that is not possible you can have a single class DomainApis routed to /domains and put the policy methods in it. On the previous example I mentioned this was the approach, the VirtualizationConnectorApis have the security group methods and they are routed under url /virtualizationconnectors/

Comment (sudhirappaji) [inline src/main/java/org/osc/manager/rest/server/api/PolicyApis.java]: OK, I prefer to handle both Domain and Policies in DomainApis like in Virtualizationconnector.
